### PR TITLE
Jmprieur/credential loader

### PR DESCRIPTION
**Pull request by jmprieur**
- Adding a service for ICredentialsLoader, so that apps can override the default implementation (which is a DefaultCertificateLoader. itself inheriting from a new class DefaultCredentialsLoader.
- Renamed ICredentialLoader to ICredentialSourceLoader, and adding more implementations (for the signed assertions from Managed identity or pod identity) 
- Moved the credentials load methods to Async,

Requires Ms.Id.Abstractions 1.0.2-preview

### CODE REVIEW

File: src/Microsoft.Identity.Web.Certificate/Base64EncodedCertificateLoader.cs [hunk 1]
```diff
@@ -4,19 +4,20 @@
...
 
 namespace Microsoft.Identity.Web
 {
-    internal sealed class Base64EncodedCertificateLoader : ICredentialLoader
+    internal sealed class Base64EncodedCertificateLoader : ICredentialSourceLoader
     {
         public CredentialSource CredentialSource => CredentialSource.Base64Encoded;
 
-        public void LoadIfNeeded(CredentialDescription credentialDescription)
+        public Task LoadIfNeededAsync(CredentialDescription credentialDescription, bool throwExceptions = false)
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificate/Base64EncodedCertificateLoader.cs:15]: >throwExceptions [](http://example.com/codeflow?start=88&length=15)

this is not used anywhere. can be removed.

File: src/Microsoft.Identity.Web.Certificate/DefaultCredentialsLoader.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            { CredentialSource.Base64Encoded, new Base64EncodedCertificateLoader() },
+            { CredentialSource.SignedAssertionFromManagedIdentity, new SignedAssertionFromManagedIdentityCredentialLoader() },
+            { CredentialSource.SignedAssertionFilePath, new SignedAssertionFilePathCredentialsLoader() },
+        };
+
+        /// <inheritdoc/>
+        /// Load the credentials from the description, if needed.
+        public async Task LoadCredentialsIfNeededAsync(CredentialDescription credentialDescription, bool throwException = false)
+        {
+            _ = Throws.IfNull(credentialDescription);
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificate/DefaultCredentialsLoader.cs:35]: >Throws [](http://example.com/codeflow?start=16&length=6)

nice :)

File: src/Microsoft.Identity.Web.Certificate/ICertificateLoader.cs [hunk 3]
```diff
@@ -21,7 +21,7 @@ namespace Microsoft.Identity.Web
     public interface ICertificateLoader
     {
         /// <summary>
-        /// Load the certificate from the description, if needed.
+        /// Load the credentials from the description, if needed.
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificate/ICertificateLoader.cs:24]: >credentials [](http://example.com/codeflow?start=21&length=11)

why change the XML if this is still relating to certs?

File: src/Microsoft.Identity.Web.Certificate/SignedAssertionFilePathCredentialsLoader.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Text;
+using Azure.Identity;
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificate/SignedAssertionFilePathCredentialsLoader.cs:5]: these usings can be removed

File: src/Microsoft.Identity.Web.Certificate/SignedAssertionFromManagedIdentityCredentialLoader.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                {
+                    // Given that managed identity can be not available locally, we need to try to get a
+                    // signed assertion, and if it fails, move to the next credentials
+                    _= await managedIdentityClientAssertion!.GetSignedAssertion(CancellationToken.None);
+                }
+                catch (AuthenticationFailedException)
+                {
+                    credentialDescription.CachedValue = managedIdentityClientAssertion;
+                    credentialDescription.Skip = true;
+                    if (throwExceptions)
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificate/SignedAssertionFromManagedIdentityCredentialLoader.cs:35]: >throwExceptions [](http://example.com/codeflow?start=24&length=15)

why is this only needed for this one and not the others?

**Comment by jennyf19** [comment]
> using System;

license info

---
Refers to: src/Microsoft.Identity.Web.Certificate/SignedAssertionFromManagedIdentityCredentialLoader.cs:1 in 749672d. [](commit_id = 749672db1c48bf6a00c69f0b0f87b74d06388fba, deletion_comment = False)

File: src/Microsoft.Identity.Web.Certificate/SignedAssertionFromManagedIdentityCredentialLoader.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Text;
+using Azure.Identity;
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.Certificate/SignedAssertionFromManagedIdentityCredentialLoader.cs:5]: remove these usings

File: src/Microsoft.Identity.Web.TokenAcquisition/ConfidentialClientApplicationBuilderExtension.cs [hunk 7]
```diff
@@ -6,6 +6,7 @@
 using System.Linq;
 using System.Security.Cryptography.X509Certificates;
 using System.Threading;
+using System.Threading.Tasks;
 using Azure.Identity;
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.TokenAcquisition/ConfidentialClientApplicationBuilderExtension.cs:10]: remove usings

File: src/Microsoft.Identity.Web.TokenAcquisition/ConfidentialClientApplicationBuilderExtension.cs [hunk 8]
```diff
@@ -18,59 +19,51 @@ internal static partial class ConfidentialClientApplicationBuilderExtension
...
+            ILogger logger,
+            ICredentialsLoader credentialsLoader)
         {
             foreach (var credential in clientCredentials)
             {
                 if (!credential.Skip)
                 {
-                    // TODO, use the credential loader
+                    string errorMessage = string.Empty;
+                    try
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.TokenAcquisition/ConfidentialClientApplicationBuilderExtension.cs:30]: This doesn't seem right..want to sync?

File: src/Microsoft.Identity.Web.TokenAcquisition/TokenAcquisition.cs [hunk 9]
```diff
@@ -82,18 +85,22 @@ class OAuthConstants
...
+            IServiceProvider serviceProvider,
+            ICredentialsLoader credentialsLoader)
         {
             _tokenCacheProvider = tokenCacheProvider;
             _httpClientFactory = new MsalAspNetCoreHttpClientFactory(httpClientFactory);
             _logger = logger;
             _serviceProvider = serviceProvider;
             _tokenAcquisitionHost = tokenAcquisitionHost;
+
+            _credentialsLoader = credentialsLoader;
```
Comment (jennyf19) [inline src/Microsoft.Identity.Web.TokenAcquisition/TokenAcquisition.cs:103]: nit remove extra space

**Review by jennyf19 — APPROVED** [review]
:shipit:

File: src/Microsoft.Identity.Web.Certificate/ICertificateLoader.cs (hunk 3 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificate/ICertificateLoader.cs:24]: good catch. Thanks!

File: src/Microsoft.Identity.Web.Certificate/SignedAssertionFilePathCredentialsLoader.cs (hunk 4 above)
Comment (jmprieur) [reply src/Microsoft.Identity.Web.Certificate/SignedAssertionFilePathCredentialsLoader.cs:5]: Done

**Pull request merged**